### PR TITLE
Hiding symbols from shared object libraries

### LINKED ISSUE #1722: [FEA] RAFT should ensure all its symbols are hidden from shared object libraries
**Is your feature request related to a problem? Please describe.**

As a user of RAFT, I would like to build a shared object library, `libA.so`, that internally uses RAFT function templates, including `__global__` function templates. 

Today, RAFT does nothing to hide the visibility of its `__global__` function templates or any other host template functions, and by default these symbols have _weak_ visibility. In short, this means if I link two dynamic libraries `A.so` and `B.so` into my application that both contain identical instantiations of a RAFT template, then the linker will discard one of the two instantiations and use only one of them. This can lead to disastrous and insidious issues like spurious silent failures. 

This issue is true of any header-only, C++ template library, but is particularly bad for CUDA C++ libraries that ship `__global__` function templates. Consider this trivial example of one of _many_ ways things can go wrong

The following code has two TUs:

- volta.cu​ compiled with sm_70​
- pascal.cu​ compiled with sm_60​

Each TU has a single function ( `volta()`​ or `pascal()​` respectively) and this function queries and prints the `ptxVersion`​ of a `kernel<void>`​ using `cudaFuncGetAttributes​`.

These TUs are linked into a program that determines the compute capability of device 0 and invokes `volta()​` or `pascal​()` accordingly.

One would expect that invoking `volta​()` would always print 70​ and invoking `pascal()​` would print 60​.

However, this is not the case. As described above, the kernel​ template has weak linkage, and so when linking the `volta.o`​ and `pascal.o`​ TUs together, the linker selects one of the instantiations of `kernel<void>`​ and discards the other.

The end result is that the program will randomly print 60 or 70 depending on which instantiation the linker picked.

```cpp
// kernel.cuh
template <typename T>
__global__ void kernel(){}

// volta.cu
#include "kernel.cuh"
void volta(){
   cudaFuncAttributes attrs;
   cudaFuncGetAttributes(&attrs, (void *)(kernel<void>));
   printf("%d\n, attrs.ptxVersion*10);
}

// pascal.cu
#include "kernel.cuh"
void pascal(){

   cudaFuncAttributes attrs;
   cudaFuncGetAttributes(&attrs, (void *)(kernel<void>));
   printf("%d\n, attrs.ptxVersion*10);

}

// main.cpp
void volta();
void pascal();
int main(){
   int compute_capability;
   cudaDeviceGetAttribute(&compute_capability, cudaDevAttrComputeCapabilityMajor, 0);
   if(compute_capability >= 70)
      volta();
   else
      pascal();
}

nvcc -c -arch=sm_70 volta.cu
nvcc -c -arch=sm_60 pascal.cu
nvcc -c main.cpp
nvcc -o test volta.o pascal.o main.o
./test
```

TL;DR:
- Given a __global__​ function, kernel​, with weak linkage (like a template)
- Instantiate kernel<T>​ in separate TUs compiled with different PTX architectures
- Link the separate TUs into a single program 
- The linker will see the two instantiations of kernel<T>​ as identical and discard one
- Invoking kernel<T>()​ results in potentially executing code you did not expect

**Describe the solution you'd like**

Luckily the solution is quite simple. 

- Every `__global__` function should be annotated as `static`
- Every host template function should be annotated as `__attribute__((visibility("hidden")))`.

This makes the symbol hidden in any resulting dynamic library. 

For convenience, you'd likely want to use a macro to wrap these like this:
```
#if defined(__CUDACC__)
#       define RAFT_KERNEL \
            static __global__
#else
#   define CUB_KERNEL
#endif
```

**Additional Context**

We've been bitten by this in Thrust/CUB several times over the years. 

Thrust/CUB also have the ability to allow users to customize the namespace in order to differentiate the symbols and avoid this problem. However, this solution is not robust. First of all, it requires every user to remember to customize the namespace. Secondly, it's possible for users to properly customize the namespace and still run afoul of the issues that can result. 

See: 
- https://github.com/NVIDIA/cub/issues/545
- https://github.com/NVIDIA/cccl/issues/166
- https://github.com/NVIDIA/cutlass/issues/1027
- https://confluence.nvidia.com/x/esayW (internal only)

### PR DESCRIPTION
**Pull request by cjnolet**
Partially addresses #1722.

### CODE REVIEW

**Comment by ahendriksen** [comment]
We should check the binary size of libraft after this change. We have a lot of duplicate kernels. When they are static they will be duplicated in each TU and linked into libraft.so. The impact might also be smaller than expected because the linker seems not to deduplicate kernels with weak linkage anyway(?)

File: cpp/include/raft/random/detail/make_regression.cuh [hunk 1]
```diff
@@ -40,7 +40,10 @@ namespace detail {
 
 /* Internal auxiliary function to help build the singular profile */
 template <typename DataT, typename IdxT>
-static __global__ void _singular_profile_kernel(DataT* out, IdxT n, DataT tail_strength, IdxT rank)
+static _RAFT_KERNEL void _singular_profile_kernel(DataT* out,
```
Comment (benfred) [inline cpp/include/raft/random/detail/make_regression.cuh:43]: Since we have `#define _RAFT_KERNEL static __global__` - doesn't this expand out to `static static __global__` ? should we drop the `static` here (and in other similar places)?

```suggestion 
_RAFT_KERNEL void _singular_profile_kernel(DataT* out,
```

Comment (cjnolet) [reply cpp/include/raft/random/detail/make_regression.cuh:43]: @benfred the whole point of this macro is that according to #1722, we need to define _all_ `__global__` to be `static __global__`. I'd prefer not to have to specify `static` explicitly when it's going to have to be specified everywhere anyways.

Comment (cjnolet) [reply cpp/include/raft/random/detail/make_regression.cuh:43]: EDIT: Don't mind me, I misread your comment. Yeah absolutely what you found here was a classic find and replace error.

Comment (benfred) [reply cpp/include/raft/random/detail/make_regression.cuh:43]: I think you misunderstood me - I'm on board with changing `__global__` to `_RAFT_KERNEL`.  

I'm just saying the `static _RAFT_KERNEL` should probably be just `_RAFT_KERNEL` (so that we're not specifying static twice)

**Comment by jrhemstad** [comment]
After discussing internally with @ahendriksen, allow me to contradict myself (again) and suggest that the macro you want is actually:
```
#if defined(__CUDACC__)
#   if !defined(__CUDACC_RDC__)
#       define _RAFT_KERNEL \
            static __global__
#   else
#       define _RAFT_KERNEL \
            __attribute__ ((visibility ("hidden"))) __global__
#   endif
#else
#   define _RAFT_KERNEL
#endif
```

This is because when compiling with `-rdc=true` then the device linker will deduplicate all the identical kernels in the same shared object library and you likely want to preserve that behavior by not marking the kernels as `static`. You still want to hide the symbols from the resulting .so though, so instead we annotate it with the `hidden` attribute.

File: cpp/include/raft/random/detail/make_regression.cuh (hunk 1 above)
Comment (cjnolet) [reply cpp/include/raft/random/detail/make_regression.cuh:43]: > I think you misunderstood me - I'm on board with changing __global__ to _RAFT_KERNEL.

I'm just saying the static _RAFT_KERNEL should probably be just _RAFT_KERNEL (so that we're not specifying static twice)

Yes, absolutely I misunderstood and on second inspection realized that. Yes you are right, I did a search and replace for `__global__` but I didn't actually check for places that already had `static __global__`.

**Comment by jrhemstad** [comment]
fyi, here's the solution we ended up using in Thrust/CUB: https://github.com/NVIDIA/cccl/pull/443
